### PR TITLE
Fleshed out implementation of UseGetterOnlyAutoPropertyAnalyzer

### PR DESCRIPTION
**Pull request by pharring**
Checks if the property is assigned anywhere outside a constructor of its containing type. If so, the setter can't be removed.

Added unit tests.

### CODE REVIEW

**Comment by pharring** [comment]
I've been thinking about this and reading John Hamby's latest document. Since the setters are private, we should be able to use a "per type" analyzer instead of "per compilation". That should also keep the memory usage down. According to John's doc, that's done by using a "symbol action":
>A symbol action is invoked once per semantic processing of a declaration of a type or type member, 

I will look into that.

**Comment by pharring** [comment]
@DustinCampbell Here you go.

**Comment by pharring** [comment]
@DustinCampbell Please review. It uses a symbol analyzer now (instead of a compilation end analyzer), so it's much more responsive and it works in CTP6.

File: Source/CSharpEssentials.Tests/UseGetterOnlyAutoPropertyAnalyzerTests.cs [hunk 1]
```diff
@@ -0,0 +1,201 @@
+﻿using CSharpEssentials.GetterOnlyAutoProperty;
+using Microsoft.CodeAnalysis.Diagnostics;
+using NUnit.Framework;
+
+namespace CSharpEssentials.Tests
+{
+    [TestFixture]
+    public class UseGetterOnlyAutoPropertyAnalyzerTests : AnalyzerTestFixture
```
Comment (DustinCampbell) [inline Source/CSharpEssentials.Tests/UseGetterOnlyAutoPropertyAnalyzerTests.cs]: A couple of tests to consider adding:

1. The diagnostic exists when there are multiple constructors that write to the property.
2. The diagnostic exists when the property's written in a constructor and read from another method.
3. The diagnostic exists when the property and the constructor writing to it are in different parts of a partial class.

**Comment by pharring** [comment]
@DustinCampbell Done.

File: Source/CSharpEssentials/GetterOnlyAutoProperty/UseGetterOnlyAutoPropertyAnalyzer.cs [hunk 2]
```diff
@@ -1,35 +1,214 @@
...
+            foreach (var property in allProperties)
+            {
+                var propertySymbol = (IPropertySymbol)property;
+                if (!propertySymbol.IsReadOnly && propertySymbol.ExplicitInterfaceImplementations.IsDefaultOrEmpty)
+                {
+                    var setMethod = propertySymbol.SetMethod;
+                    if (setMethod != null && setMethod.DeclaredAccessibility == Accessibility.Private)
+                    {
+                        // Find the syntax for the setter.
+                        var declaration = setMethod.DeclaringSyntaxReferences.FirstOrDefault().GetSyntax(cancellationToken) as AccessorDeclarationSyntax;
```
Comment (DustinCampbell) [inline Source/CSharpEssentials/GetterOnlyAutoProperty/UseGetterOnlyAutoPropertyAnalyzer.cs]: Can FirstOrDefault() throw here?

File: Source/CSharpEssentials/GetterOnlyAutoProperty/UseGetterOnlyAutoPropertyAnalyzer.cs [hunk 3]
```diff
@@ -1,35 +1,214 @@
...
+
+            return false;
+        }
+
+        private static Diagnostic CreateDiagnostic(ISymbol symbol, CancellationToken cancellationToken)
+        {
+            foreach (var reference in symbol.DeclaringSyntaxReferences)
+            {
+                // The span should be on the setter, but the symbol is for the whole property declaration.
+                var declarationNode = reference.GetSyntax(cancellationToken) as PropertyDeclarationSyntax;
```
Comment (DustinCampbell) [inline Source/CSharpEssentials/GetterOnlyAutoProperty/UseGetterOnlyAutoPropertyAnalyzer.cs]: "as" cast without testing without testing for null. If we're confident that's a PropertyDeclarationSyntax, maybe just a direct cast?

**Comment by DustinCampbell** [comment]
Other than the various nits. This looks great to me! Obviously, a SymbolAnalyzer isn't ideal since it may force the compiler to do some work twice, but it's better than using CompilationEnded performance-wise.

File: Source/CSharpEssentials.Tests/AnalyzerTestFixture.cs [hunk 4]
```diff
@@ -18,7 +18,7 @@ protected void NoDiagnostic(string code, string diagnosticId)
 
             var diagnostics = GetDiagnostics(document);
 
-            Assert.That(diagnostics.All(d => d.Id == diagnosticId), Is.True);
+            Assert.That(diagnostics.Any(d => d.Id == diagnosticId), Is.False);
```
Comment (DustinCampbell) [inline Source/CSharpEssentials.Tests/AnalyzerTestFixture.cs]: Oops. :smile:

File: Source/CSharpEssentials/GetterOnlyAutoProperty/UseGetterOnlyAutoPropertyAnalyzer.cs [hunk 5]
```diff
@@ -1,35 +1,214 @@
...
+                node = node.Parent;
+            }
 
-            if (accessorDeclaration.Body == null &&
-                !accessorDeclaration.SemicolonToken.IsMissing &&
-                accessorDeclaration.Modifiers.Any(SyntaxKind.PrivateKeyword))
+            return false;
+        }
+
+        private static bool IsUpdatingExpression(ref SyntaxNode node)
```
Comment (DustinCampbell) [inline Source/CSharpEssentials/GetterOnlyAutoProperty/UseGetterOnlyAutoPropertyAnalyzer.cs]: Based on our chat, this isn't quite right. It'll return true for this case:

```C#
int x = 0;
x += P;
```

File: Source/CSharpEssentials/GetterOnlyAutoProperty/UseGetterOnlyAutoPropertyAnalyzer.cs [hunk 6]
```diff
@@ -1,35 +1,214 @@
...
+        {
+            var identifier = node;
+            for (node = node.Parent; node != null; node = node.Parent)
             {
-                context.ReportDiagnostic(
-                    Diagnostic.Create(
-                        DiagnosticDescriptors.UseGetterOnlyAutoProperty,
-                        accessorDeclaration.GetLocation()));
+                switch (node.Kind())
```
Comment (DustinCampbell) [inline Source/CSharpEssentials/GetterOnlyAutoProperty/UseGetterOnlyAutoPropertyAnalyzer.cs]: For the simple and compound assignment, you should do the same thing, which is to bind the Left value of the assignment and verify that it is the same property symbol.

**Comment by pharring** [comment]
@DustinCampbell One more time. I think I addressed everything.

File: Source/CSharpEssentials/GetterOnlyAutoProperty/UseGetterOnlyAutoPropertyAnalyzer.cs (hunk 6 above)
Comment (pharring) [inline Source/CSharpEssentials/GetterOnlyAutoProperty/UseGetterOnlyAutoPropertyAnalyzer.cs]: Done

File: Source/CSharpEssentials/GetterOnlyAutoProperty/UseGetterOnlyAutoPropertyAnalyzer.cs (hunk 2 above)
Comment (pharring) [inline Source/CSharpEssentials/GetterOnlyAutoProperty/UseGetterOnlyAutoPropertyAnalyzer.cs]: Done

**Comment by DustinCampbell** [comment]
Looks great!

**Pull request merged**